fuzzing_cli/fuzz/ide/hardhat.py

```diff
@@ -24,3 +31,132 @@ def get_default_build_dir() -> Path:
     @staticmethod
     def get_default_sources_dir() -> Path:
         return Path.cwd().joinpath("contracts")
+
+    def _get_build_info(self, build_dir: Path) -> Dict[str, Dict[str, Path]]:
+        if not build_dir.is_dir():
+            raise BuildArtifactsError("Build directory doesn't exist")
+
+        # Here we need to resolve sources_dir relative to build_dir's parent
+        # because sources_dir can be provided as a relative or an absolute path deeply nested
+        # in the sources directory (which is `contracts` by default in Hardhat), and we need to
+        # resolve it to the same level as build_dir, because Hardhat stores build artifacts
+        # in the `artifacts` directory which is at the same level as the `contracts` directory
+        built_contracts = build_dir.joinpath(
+            self.sources_dir.relative_to(build_dir.parent)
+        )
+
+        result = defaultdict(dict)
+
+        for child in built_contracts.glob("**/*.dbg.json"):
+            if not child.is_file():
+                continue
+
+            if child.name.startswith(
+                "."
+            ):  # some hidden file (probably created by OS, especially the macOS)
+                continue
+
+            source_file_rel_path = str(child.relative_to(build_dir).parent)
+            contract_name = child.name.split(".dbg.json")[0]
+
+            with open(child, "r") as f:
+                try:
+                    dbg_info = json.load(f)
+                except Exception:
+                    continue
+                # buildInfo is relative to the artifacts directory (with symlinks, like ../),
+                # so we need to join the path with the build_dir and resolve it
+                result[source_file_rel_path][contract_name] = child.parent.joinpath(
+                    Path(dbg_info["buildInfo"])
+                ).resolve()
+
+        return result
+
+    @property
+    @lru_cache(maxsize=1)
+    def build_info(self) -> Dict[str, Dict[str, Path]]:
+        return self._get_build_info(self.build_dir)
+
+    def get_source(self, source_path: str, sources: Dict[str, Dict[str, str]]) -> str:
+        if (
+            self.map_to_original_source
+            and Path(self.normalize_path(source_path) + ".original").is_file()
+        ):
+            return get_content_from_file(self.normalize_path(source_path) + ".original")
+        return sources[source_path]["content"]
+
+    @lru_cache(maxsize=1)
+    def process_artifacts(self) -> Tuple[Dict[str, List[Contract]], Dict[str, Source]]:
+        result_contracts = defaultdict(list)
+        result_sources = {}
+
+        build_info_paths: Dict[Path, Dict[str, List[str]]] = defaultdict(
+            partial(defaultdict, list)
+        )
+        for source_file_name, contracts in self.build_info.items():
```

Comment:
Can we add code here that:

1. Detects if the same `source_file_name` appears in multiple build_info files
2. Communicates a warning/error to the FaaS API when this happens

I want to be able to detect when this happens in practice so we can debug it and adapt the code